src/java/org/apache/cassandra/index/sai/disk/v1/bbtree/BlockBalancedTreeWriter.java

```diff
@@ -0,0 +1,766 @@
+ *                                                  | LENGTH | BYTES   |
+ *                                                  +--------+---------+
+ *  </pre>
+ *
+ * <p>
+ * <b>NOTE</b>: This can write at most Integer.MAX_VALUE * <code>maxPointsInLeafNode</code> total points.
+ * <p>
+ * @see BKDWriter
+ */
+@NotThreadSafe
+public class BlockBalancedTreeWriter
+{
+    // Enable to check that values are added to the tree in correct order and within bounds
+    public static final boolean DEBUG = CassandraRelevantProperties.SAI_TEST_BALANCED_TREE_DEBUG_ENABLED.getBoolean();
+
+    // Default maximum number of point in each leaf block
+    public static final int DEFAULT_MAX_POINTS_IN_LEAF_NODE = 1024;
+
+    private final int bytesPerValue;
+    private final int maxPointsInLeafNode;
+    private final byte[] minPackedValue;
+    private final byte[] maxPackedValue;
+    private long valueCount;
+
+    public BlockBalancedTreeWriter(int bytesPerValue, int maxPointsInLeafNode)
+    {
+        if (maxPointsInLeafNode <= 0)
+            throw new IllegalArgumentException("maxPointsInLeafNode must be > 0; got " + maxPointsInLeafNode);
+        if (maxPointsInLeafNode > ArrayUtil.MAX_ARRAY_LENGTH)
+            throw new IllegalArgumentException("maxPointsInLeafNode must be <= ArrayUtil.MAX_ARRAY_LENGTH (= " +
+                                               ArrayUtil.MAX_ARRAY_LENGTH + "); got " + maxPointsInLeafNode);
+
+        this.maxPointsInLeafNode = maxPointsInLeafNode;
+        this.bytesPerValue = bytesPerValue;
+
+        minPackedValue = new byte[bytesPerValue];
+        maxPackedValue = new byte[bytesPerValue];
+    }
+
+    public long getValueCount()
+    {
+        return valueCount;
+    }
+
+    public int getBytesPerValue()
+    {
+        return bytesPerValue;
+    }
+
+    public int getMaxPointsInLeafNode()
+    {
+        return maxPointsInLeafNode;
+    }
+
+    /**
+     * Write the sorted values from a {@link BlockBalancedTreeIterator}.
+     * <p>
+     * @param treeOutput The {@link IndexOutput} to write the balanced tree to
+     * @param values The {@link BlockBalancedTreeIterator} containing the values and rowIDs to be written
+     * @param callback The {@link Callback} used to record the leaf postings for each leaf
+     *
+     * @return The file pointer to the beginning of the balanced tree
+     */
+    public long write(IndexOutput treeOutput, BlockBalancedTreeIterator values, final Callback callback) throws IOException
+    {
+        SAICodecUtils.writeHeader(treeOutput);
+
+        LeafWriter leafWriter = new LeafWriter(treeOutput, callback);
+
+        while (values.hasNext())
+        {
+            Pair<byte[], PostingList> pair = values.next();
+            while (true)
+            {
+                long segmentRowId = pair.right.nextPosting();
+                if (segmentRowId == END_OF_STREAM)
+                    break;
+                leafWriter.add(pair.left, segmentRowId);
+            }
```

Comment:
nit: Alternate version w/ no unconditional loop:

```
long segmentRowId = pair.right.nextPosting();
while (segmentRowId != END_OF_STREAM)
{
    leafWriter.add(pair.left, segmentRowId);
    segmentRowId = pair.right.nextPosting();
}

Reply:
I almost thought we could make `PostingList` `Iterable`, but then though maybe avoiding possible boxing there would be best.

Reply:
We should definitely avoid boxing the values but I have changed the loop.